lib/reply.js

```diff
@@ -157,66 +142,91 @@ function onSendEnd (reply, payload) {
-    flatstr(payload)
-  } else if (reply._serializer) { // All the code below must have a 'content-type' setted
-    payload = reply._serializer(payload)
+    payload = serialize(reply.context, payload, reply.res.statusCode)
   }
 
+  flatstr(payload)
+
   if (!reply.res.getHeader('Content-Length')) {
     reply.res.setHeader('Content-Length', '' + Buffer.byteLength(payload))
   }
   reply.sent = true
   reply.res.end(payload)
 }
 
-function handleError (reply, err, cb) {
-  if (!reply.res.statusCode || reply.res.statusCode < 400) {
-    reply.res.statusCode =
-        (err === null || err === undefined) ? 500
-      : (err.status >= 400) ? err.status
-      : (err.statusCode >= 400) ? err.statusCode
-      : 500
+function handleError (reply, error, cb) {
+  var statusCode = reply.res.statusCode
+  if (error == null) {
+    statusCode = statusCode || 500
+  } else if (error.status >= 400) {
+    statusCode = error.status
+  } else if (error.statusCode >= 400) {
+    statusCode = error.statusCode
+  } else {
+    statusCode = statusCode || 500
   }
+  if (statusCode < 400) statusCode = 500
+  reply.res.statusCode = statusCode
 
-  if (reply.res.statusCode >= 500) {
-    reply.res.log.error({ res: reply.res, err }, err && err.message)
-  } else if (reply.res.statusCode >= 400) {
-    reply.res.log.info({ res: reply.res, err }, err && err.message)
+  if (statusCode >= 500) {
+    reply.res.log.error({ res: reply.res, error }, error && error.message)
+  } else if (statusCode >= 400) {
+    reply.res.log.info({ res: reply.res, error }, error && error.message)
   }
 
-  var errorHandler = reply.context.errorHandler
+  if (error && error.headers) {
+    reply.headers(error.headers)
+  }
 
-  if (errorHandler && !reply._errored) {
+  var customErrorHandler = reply.context.errorHandler
+  if (customErrorHandler && reply._customError === false) {
     reply.sent = false
-    reply._errored = true
-    errorHandler(err, reply)
+    reply._isError = false
+    reply._customError = true
+    var result = customErrorHandler(error, reply)
+    if (result && typeof result.then === 'function') {
+      result.then(payload => reply.send(payload))
+            .catch(err => reply.send(err))
+    }
     return
   }
 
-  reply.res.setHeader('Content-Type', 'application/json')
+  var payload = {
+    error: statusCodes[statusCode + ''],
+    message: error ? error.message : '',
+    statusCode: statusCode
+  }
 
-  cb(reply, Object.assign({
-    error: statusCodes[reply.res.statusCode + ''],
-    message: (err || '') && err.message,
-    statusCode: reply.res.statusCode
-  }, reply._extendServerError && reply._extendServerError(err)))
+  if (cb) {
+    cb(reply, payload)
+    return
+  }
+
+  payload = serializeError(payload)
+  flatstr(payload)
+
+  reply.res.setHeader('Content-Length', '' + Buffer.byteLength(payload))
+  reply.res.setHeader('Content-Type', 'application/json')
+  reply.sent = true
+  reply.res.end(payload)
```

Comment:
What happen if I set the custom serializer ?

Reply:
The custom serializer works for the normal cases, the error object produced by Fastify is always the same unless you use the `setCustomError` function (but in that case it will not be handled by this function).